akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -334,3 +339,85 @@ private[akka] final case class Expand[In, Out, Seed](seed: In ⇒ Seed, extrapol
   final override def restart(): Expand[In, Out, Seed] =
     throw new UnsupportedOperationException("Expand doesn't support restart")
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Future[Out], decider: Supervision.Decider)
+  extends AsyncStage[In, Out, (Int, Try[Out])] {
+
+  type Notification = (Int, Try[Out])
+
+  private var callback: AsyncCallback[Notification] = _
+  private val elemsInFlight = FixedSizeBuffer[Try[Out]](parallelism)
+
+  override def initAsyncInput(ctx: AsyncContext[Out, Notification]): Unit = {
+    callback = ctx.getAsyncCallback()
+  }
+
+  override def onPush(elem: In, ctx: AsyncContext[Out, Notification]) = {
+    val future =
+      try f(elem)
+      catch {
+        case NonFatal(ex) if decider(ex) != Supervision.Stop ⇒ null // resume processing
+      }
+    if (future ne null) {
+      val idx = elemsInFlight.enqueue(null)
```

Comment:
this reads weird, but I see what it does. Maybe it would help if you would preallocate a dummy Try[Out] (obviously a dummy failure will do), naming it WaitingSlot or something and using it instead of null. That would make the intent more clear.